File: app/src/main/java/com/github/bjoernpetersen/q/QueueState.java
package com.github.bjoernpetersen.q;

import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry;
import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public final class QueueState {

  private static final String TAG = QueueState.class.getSimpleName();
  private static final QueueState INSTANCE = new QueueState();

  private final Set<WeakListener> listeners;
  private List<QueueEntry> queue;

  private QueueState() {
    this.listeners = new HashSet<>();
    this.queue = Collections.emptyList();
  }

  public static QueueState getInstance() {
    return INSTANCE;
  }

  public List<QueueEntry> get() {
    return queue;
  }

  public void set(List<QueueEntry> queue) {
    if (queue == null) {
      throw new NullPointerException();
    }
    List<QueueEntry> oldQueue = this.queue;
    this.queue = Collections.unmodifiableList(queue);
    for (WeakListener listener : listeners) {
      listener.onChange(oldQueue, this.queue);
    }
  }

  public void addListener(Listener listener) {
    listeners.add(new WeakListener(listener));
  }

  public void removeListener(Listener listener) {
    listeners.remove(new WeakListener(listener));
  }

  private final class WeakListener implements Listener {

    private final WeakReference<Listener> listener;

    WeakListener(Listener listener) {
      this.listener = new WeakReference<>(listener);
    }

    @Override
    public void onChange(List<QueueEntry> oldQueue, List<QueueEntry> newQueue) {
      Listener listener = this.listener.get();
      if (listener == null) {
        listeners.remove(this);
      } else {
        listener.onChange(oldQueue, newQueue);
      }
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }

      WeakListener other = (WeakListener) o;
      Listener otherListener = other.listener.get();
      Listener listener = this.listener.get();
      if (listener == null) {
        return otherListener == null;
      }
      return listener.equals(otherListener);
    }

    @Override
    public int hashCode() {
      Listener listener = this.listener.get();
      return listener == null ? 0 : listener.hashCode();
    }
  }

  public interface Listener {

    void onChange(List<QueueEntry> oldQueue, List<QueueEntry> newQueue);
  }
}

File: app/src/main/java/com/github/bjoernpetersen/q/QueueState.kt
package com.github.bjoernpetersen.q

import android.os.Looper
import android.support.annotation.UiThread
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry
import java.lang.ref.WeakReference
import java.util.*

@UiThread
object QueueState {

  private val listeners: MutableSet<WeakListener> = HashSet()
  var queue: List<QueueEntry> = emptyList()
    @UiThread
    set(value) {
      if (Looper.getMainLooper().thread != Thread.currentThread()) {
        throw RuntimeException("Must be called on UI thread")
      }
      val oldQueue = this.queue
      field = value
      for (listener in listeners) {
        listener(oldQueue, value)
      }
    }

  @UiThread
  fun addListener(listener: (List<QueueEntry>, List<QueueEntry>) -> Unit) {
    listeners.add(WeakListener(listener))
  }

  @UiThread
  fun removeListener(listener: (List<QueueEntry>, List<QueueEntry>) -> Unit) {
    listeners.remove(WeakListener(listener))
  }

  private class WeakListener(listener: (List<QueueEntry>, List<QueueEntry>) -> Unit) :
      (List<QueueEntry>, List<QueueEntry>) -> Unit {

    private val listener: WeakReference<(List<QueueEntry>, List<QueueEntry>) -> Unit> = WeakReference(
        listener)

    override fun invoke(oldQueue: List<QueueEntry>, newQueue: List<QueueEntry>) {
      val listener = this.listener.get()
      listener?.invoke(oldQueue, newQueue) ?: listeners.remove(this)
    }

    override fun equals(other: Any?): Boolean {
      if (this === other) {
        return true
      }
      if (other == null || javaClass != other.javaClass) {
        return false
      }

      other as WeakListener
      return other.listener.get() == this.listener.get()
    }

    override fun hashCode(): Int {
      val listener = this.listener.get()
      return listener?.hashCode() ?: 0
    }
  }
}
